### PR TITLE
Concurrency fix

### PR DESCRIPTION
**Pull request by davidhadas**
1. Protect Pile in guard-gate as it may be concurrently updated

1. Protect Pile in guard-service as it may be concurrently updated

1. Protect the  guard-service cache and namespace maps from concurrency read/write

### CODE REVIEW

File: cmd/guard-service/services.go [hunk 1]
```diff
@@ -64,65 +68,91 @@ func (s *services) start() {
...
+
+	s.mutex.Lock()
+	// check if known Namespace
+	_, knownNamespace = s.namespaces[ns]
+	if !knownNamespace {
+		s.namespaces[ns] = true
+	}
 	// try to get from cache
 	record := s.cache[service]
+	s.mutex.Unlock()
```
Comment (psschwei) [inline cmd/guard-service/services.go:134]: `defer`?

File: cmd/guard-service/services.go [hunk 2]
```diff
@@ -132,12 +162,11 @@ func (s *services) set(ns string, sid string, cmFlag bool, guardianSpec *spec.Gu
 		record.cmFlag = cmFlag
 		s.cache[service] = record
 	}
+	s.mutex.Unlock()
```
Comment (psschwei) [inline cmd/guard-service/services.go:165]: `defer`

File: cmd/guard-service/services.go [hunk 3]
```diff
@@ -152,7 +181,9 @@ func (s *services) update(ns string, sid string, cmFlag bool, guardianSpec *spec
 
 // update the record pile by merging a new pile
 func (s *services) merge(record *serviceRecord, pile *spec.SessionDataPile) {
+	record.pileMutex.Lock()
 	record.pile.Merge(pile)
+	record.pileMutex.Unlock()
```
Comment (psschwei) [inline cmd/guard-service/services.go:186]: `defer`

File: cmd/guard-service/services.go [hunk 4]
```diff
@@ -167,9 +198,11 @@ func (s *services) learnPile(record *serviceRecord) bool {
 	}
 	config := new(spec.SessionDataConfig)
 
-	// TBD move to periodical under Ticker
+	record.pileMutex.Lock()
 	config.Learn(&record.pile)
 	record.pile.Clear()
+	record.pileMutex.Unlock()
```
Comment (psschwei) [inline cmd/guard-service/services.go:204]: `defer`

File: pkg/guard-gate/client.go [hunk 5]
```diff
@@ -127,7 +132,11 @@ func (srv *gateClient) reportPile() {
 
 	srv.httpClient.ReadToken(guardKubeMgr.ServiceAudience)
 
+	// protect pile internals read/write
+	srv.pileMutex.Lock()
 	postBody, marshalErr := json.Marshal(srv.pile)
+	srv.pileMutex.Unlock()
```
Comment (psschwei) [inline pkg/guard-gate/client.go:138]: `defer`

File: pkg/guard-gate/client.go [hunk 6]
```diff
@@ -167,12 +176,18 @@ func (srv *gateClient) reportPile() {
 }
 
 func (srv *gateClient) addToPile(profile *spec.SessionDataProfile) {
+	// protect pile internals read/write
+	srv.pileMutex.Lock()
 	srv.pile.Add(profile)
+	srv.pileMutex.Unlock()
```
Comment (psschwei) [inline pkg/guard-gate/client.go:182]: `defer`

File: pkg/guard-gate/client.go [hunk 7]
```diff
@@ -167,12 +176,18 @@ func (srv *gateClient) reportPile() {
...
+
 	pi.Log.Debugf("Learn - add to pile! pileCount %d", srv.pile.Count)
 }
 
 func (srv *gateClient) clearPile() {
+	srv.pileMutex.Lock()
 	srv.pile.Clear()
+	srv.pileMutex.Unlock()
```
Comment (psschwei) [inline pkg/guard-gate/client.go:190]: `defer`

File: pkg/guard-utils/stats.go [hunk 8]
```diff
@@ -18,20 +18,28 @@ package guardutils
...
 
 func (s *Stat) Init() {
+	s.mutex = new(sync.Mutex)
 	s.statistics = make(map[string]uint32, 8)
 }
 
 func (s *Stat) Add(key string) {
+	s.mutex.Lock()
 	s.statistics[key]++
+	s.mutex.Unlock()
```
Comment (psschwei) [inline pkg/guard-utils/stats.go:37]: `defer`

File: pkg/guard-utils/stats.go [hunk 9]
```diff
@@ -18,20 +18,28 @@ package guardutils
...
 }
 
 func (s *Stat) Log() string {
-	return fmt.Sprintf("%v", s.statistics)
+	s.mutex.Lock()
+	str := fmt.Sprintf("%v", s.statistics)
+	s.mutex.Unlock()
```
Comment (psschwei) [inline pkg/guard-utils/stats.go:43]: `defer`

**Comment by davidhadas** [comment]
There are some cases where we can simply change to defer without consequences, 
But in other cases this introduces one of the following potential problems: 
1. Using defer will result in cases where the mutex is taken twice
2. Using defer will result in (relatively) long  processing with the mutex long for no reason

So, I suggest we use the idiom only where none of the 2 occur.

**Review by psschwei — COMMENTED** [review]
/lgtm

**Pull request merged**